Labchecker 1.0

Yikun opened the pull request:
```
Usage: labchecker [--type <plugin-type>] [--cloud <cloud-provider-name>] [--nocolor]
optional arguments:
	--help		Show help message and exit.
	--type		Specify a plugin type, like 'provider', 'all'(default)
	--cloud		Specify a cloud provider, like 'otc', 'vexxhost', 'all'(default).
	--nocolor	Enable the no color mode.
	--recover	Enable the auto recover mode.
Exmaple:
labchecker check --type all --cloud all
labchecker check --type nodepool --cloud vexxhost
labchecker check --type nodepool --cloud vexxhost --recover
labchecker check --type nodepool --cloud vexxhost --nocolor

```

### Status check
![image](https://user-images.githubusercontent.com/1736354/54986686-14209080-4fee-11e9-948a-3e553ac294cb.png)

### nocolor mode
![image](https://user-images.githubusercontent.com/1736354/54986711-1be03500-4fee-11e9-80fd-d6de86ccd54d.png)

### Auto recover:
![image](https://user-images.githubusercontent.com/1736354/54986748-30bcc880-4fee-11e9-9c4c-aa8f880be65f.png)

Yikun committed:
Labchecker 1.0

Yikun commented on labchecker/labcheck.py:
```diff
@@ -1,25 +1,48 @@
 def main(argv):
     cloud = 'all'
-    type = ''
+    ptype = ''
```
@wangxiyuan type has been changed to ptype to avoid using the build-in keywords.

wangxiyuan replied:
cool

Yikun commented on labchecker/utils.py:
```diff
@@ -0,0 +1,19 @@
+def _color(s, color='b'):
+    global NOCOLOR
+    if NOCOLOR:
+        return s
```
@liu-sheng nocolor mode has been added.

Yikun commented on labchecker/plugins/recover.py:
```diff
@@ -0,0 +1,48 @@
+    NETWORK_SUBNET_CIDR = 7
+
+
+RECOVER_MAPS = {
```
@wangxiyuan it has been changed to capital letter.

wangxiyuan replied:
++，BTW，Instead collect all recover functions here, I think move  each one to related plugin is better. It could make the code  more readable and decouple.

Any way, the pr looks good enough.

Yikun reviewed (commented):
The Labkeeper repo has been rebuilt, the original pr has been deleted, so I re-post a new pr in here. and @liu-sheng @wangxiyuan comments has been addressed.

wangxiyuan commented on labchecker/labcheck.py:
```diff
@@ -32,20 +55,31 @@ def main(argv):
-            type = a
+            ptype = a
+        elif o in ("n", "--nocolor"):
+            setnocolor()
```
just use `utils.NOCOLOR=True`

wangxiyuan commented on labchecker/labcheck.py:
```diff
@@ -32,20 +55,31 @@ def main(argv):
+        header = "%s/%s. %s cloud check" % (i+1, cnt, c)
+        header_print(header)
+        for plugin in plugins:
+                plugin.cloud = c
```
again, indention problem

Yikun committed:
Fix indention and global var

kiwik commented on labchecker/labcheck.py:
```diff
@@ -1,25 +1,49 @@
 
 def usage():
-    print("Usage: labchecker [--type <plugin-type>] [--cloud <cloud-provider-name>]")
+    print("Usage: labchecker [--type <plugin-type>] [--cloud <cloud-provider-name>] [--nocolor]")
```
Please add an option to control whether tool recover cloud setting.

kiwik commented on labchecker/labcheck.py:
```diff
@@ -1,25 +1,49 @@
+    print("Usage: labchecker [--type <plugin-type>] [--cloud <cloud-provider-name>] [--nocolor]")
     print("optional arguments:")
     print("\t--help\t\tShow help message and exit.")
     print("\t--type\t\tSpecify a plugin type, like 'provider', 'all'(default)")
```
please update help message, type is `provider` or `jobs` or `nodepool` now, right? what's the different with provider and jobs?

Yikun replied:
Current, only `nodepool` and `jobs` is available.
- `nodepool` is for checking the nodepool side status, like the provider status check, nodepool configuration check
- `jobs` is for checking the zuul-jobs, or some custom jobs requirement.

and the description is out-of-date, I will update in next ps

kiwik commented on labchecker/plugins/nodepool/network.py:
```diff
@@ -0,0 +1,32 @@
+    def check(self):
+        self.failed = False
+        self.reasons = []
+        net = 'openstack --os-cloud %s network list -f value -c Name | grep openlab-net' % self.cloud
```
1. Why do not use `network show openlab-net`? it's faster
2. If there are more than one `openlab-net`, it's wrong cloud setting too.

Yikun replied:
1. Yes, I will update in next PS
2. ++

kiwik commented on labchecker/plugins/nodepool/securitygroup.py:
```diff
@@ -0,0 +1,37 @@
+    def check(self):
+        self.failed = False
+        self.reasons = []
+        sg = 'openstack --os-cloud %s security group list -f value -c Name | grep openlab-sg' % self.cloud
```
ditto

Yikun committed:
Add --recover mode

only do auto recover when the admins specify the --recover args.

kiwik commented on labchecker/plugins/jobs/image.py:
```diff
@@ -0,0 +1,19 @@
+        self.reasons = []
+        # find 'cirros' in openstack image list
+        image_check = 'openstack --os-cloud %s image list ' \
+                      '-c "Name" -c "Status" -f value | grep cirros' % self.cloud
```
There are public cirros images in some public clouds, so we should use more special name for it, and update zuul jobs to apply it.

Yikun commented:
@kiwik the `--recover` mode has been added in 68fac2a

kiwik commented on labchecker/plugins/nodepool/flavor.py:
```diff
@@ -0,0 +1,25 @@
+    def check(self):
+        self.failed = False
+        self.reasons = []
+        sg = 'openstack --os-cloud %s flavor list -f yaml -c VCPUs -c RAM -c Name' % self.cloud
```
I think we should output `DISK` size of the flavor we find, that impact whether we use boot from image or volume instance.

Yikun replied:
```
[nodepool] flavor----------------------- PASSED
Info:
- Flavor 4U8G: not found
- Flavor 8U8G: ['v2-highcpu-8 (0G)', 'v1-standard-8 (200G)']
- Flavor 1U2G: not found
- Flavor 2U2G: ['v2-highcpu-2 (0G)']
```
ok, will add disk info after flavor name, like above

Yikun committed:
Fix help usage and add recover hint

- Add recover hint of error
- Append disk info of flavor
- Optimize the network and sg check
- Fix out-of-date help usage
- Remove unused provider plugins

kiwik reviewed (approved):
LGTM, @Yikun thank you addressing my comments.

wangxiyuan commented on labchecker/labcheck.py:
```diff
@@ -32,20 +59,33 @@ def main(argv):
+
+    if ptype:
+        # Filter the plugins with specific ptype
+        plugins = list(filter(lambda x: x.ptype == ptype,
```
good to support `ptype` as a list. Can be done in a follow-up pr if needed.

wangxiyuan reviewed (commented):
LGTM.

liusheng reviewed (commented):
Just have a question didn't asked in you demo in the meeting :)

for now we can only perform checking on the nodepool node, right ? maybe we should also support checking on different hosts ? on zuul hosts or in HA deployment mode, maybe in other cloud. I mean we may need add more checks in the future, such as zuul services health status check. so maybe it is better to support run checks on different nodes by using ansible mechanism.

Also, maybe you can consider to implement these python code in ansible `libary`, Ansible can also directly using python plugin implement.

Just my personal thoughts.

Yikun commented:
> for now we can only perform checking on the nodepool node, right ? 

Yes or No. 
No is that the only config we needed is `/etc/openstack/cloud.yaml`, that is all checks is just only for **cloud providers**.
Yes is that only nodepool had this `/etc/openstack/cloud.yaml`.

> maybe we should also support checking on different hosts ?

In future, we should add check for nodepool host and zuul host to check some host status or config.


> on zuul hosts or in HA deployment mode, maybe in other cloud. I mean we may need add more checks in the future, such as zuul services health status check. so maybe it is better to support run checks on different nodes by using ansible mechanism.
> 
> Also, maybe you can consider to implement these python code in ansible `libary`, Ansible can also directly using python plugin implement.
> 
> Just my personal thoughts.

Good idea.

liusheng reviewed (approved):
let's make the fist version landed firstly, and we can improve in next steps!
